From: Jeremy Bicha <jbicha@debian.org>
Date: Sat, 9 Feb 2019 16:32:03 -0500
Subject: sound: Allow volume to be set above 100%.

Some systems have low maximum volume set (like x220), allow, from an option
in gnome-control-center to set it above that 100% limit from g-s-d
(keyboard) and gnome-shell.
Only show the above 100% volume option if:
1. you are in an ubuntu session
2. the selected output supports amplified volume. If so:
   present the settings to google that on and off. It will enable
   GNOME Shell and media keys to set the sound above 100%. If not
   enabled, volume and sliders are all capped to 100%. (LP: #1706524)
Modified from original patch in unity-control-center from Lars Uebernickel.
Adapted for Ubuntu gnome-control-center by Didier Roche
Updated for GNOME 3.32 by Jeremy Bicha
Origin: ubuntu
Bug-Ubuntu: https://launchpad.net/bugs/1706524
Bug: https://bugzilla.gnome.org/show_bug.cgi?id=710424
---
 panels/sound/cc-sound-panel.c  | 14 ++++++++++++++
 panels/sound/cc-sound-panel.ui | 10 ++++++++++
 2 files changed, 24 insertions(+)

diff --git a/panels/sound/cc-sound-panel.c b/panels/sound/cc-sound-panel.c
index 219e6c2..3c09d1d 100644
--- a/panels/sound/cc-sound-panel.c
+++ b/panels/sound/cc-sound-panel.c
@@ -34,6 +34,7 @@
 #include "cc-device-combo-box.h"
 #include "cc-fade-slider.h"
 #include "cc-level-bar.h"
+#include "cc-list-row.h"
 #include "cc-output-test-window.h"
 #include "cc-profile-combo-box.h"
 #include "cc-sound-panel.h"
@@ -53,6 +54,7 @@ struct _CcSoundPanel
   CcProfileComboBox   *output_profile_combo_box;
   CcVolumeSlider      *output_volume_slider;
   CcBalanceSlider     *balance_slider;
+  CcListRow           *allow_amplify_row;
   AdwPreferencesRow   *fade_row;
   CcFadeSlider        *fade_slider;
   AdwPreferencesRow   *subwoofer_row;
@@ -299,6 +301,7 @@ cc_sound_panel_class_init (CcSoundPanelClass *klass)
   gtk_widget_class_bind_template_child (widget_class, CcSoundPanel, output_profile_row);
   gtk_widget_class_bind_template_child (widget_class, CcSoundPanel, output_profile_combo_box);
   gtk_widget_class_bind_template_child (widget_class, CcSoundPanel, output_volume_slider);
+  gtk_widget_class_bind_template_child (widget_class, CcSoundPanel, allow_amplify_row);
   gtk_widget_class_bind_template_child (widget_class, CcSoundPanel, balance_slider);
   gtk_widget_class_bind_template_child (widget_class, CcSoundPanel, fade_row);
   gtk_widget_class_bind_template_child (widget_class, CcSoundPanel, fade_slider);
@@ -344,6 +347,17 @@ cc_sound_panel_init (CcSoundPanel *self)
                            G_CONNECT_SWAPPED);
   allow_amplified_changed_cb (self);
 
+  const gchar *desktop_list = g_getenv ("XDG_CURRENT_DESKTOP");
+  g_auto(GStrv) desktops = NULL;
+  if (desktop_list)
+    desktops = g_strsplit (desktop_list, ":", 0);
+  if (desktops && g_strv_contains ((const gchar * const *) desktops, "ubuntu")) {
+    g_settings_bind (self->sound_settings, "allow-volume-above-100-percent",
+                     self->allow_amplify_row, "active", G_SETTINGS_BIND_DEFAULT);
+  } else {
+    gtk_widget_set_visible (GTK_WIDGET (self->allow_amplify_row), FALSE);
+  }
+
   self->mixer_control = gvc_mixer_control_new ("GNOME Settings");
   gvc_mixer_control_open (self->mixer_control);
 
diff --git a/panels/sound/cc-sound-panel.ui b/panels/sound/cc-sound-panel.ui
index 0d6b12c..e3a52dd 100644
--- a/panels/sound/cc-sound-panel.ui
+++ b/panels/sound/cc-sound-panel.ui
@@ -200,6 +200,16 @@
                 </child>
               </object>
             </child>
+            <!-- Over-Amplification -->
+            <child>
+              <object class="CcListRow" id="allow_amplify_row">
+                <property name="title" translatable="yes">Over-Amplification</property>
+                <property name="subtitle" translatable="yes">Allows raising the volume above 100%. This can result in a loss of audio quality; it is better to increase application volume settings, if possible.</property>
+                <property name="subtitle-lines">6</property>
+                <property name="use-underline">True</property>
+                <property name="show-switch">True</property>
+              </object>
+            </child>
           </object>
         </child>
         <child>
